bentoml/_internal/resource.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
+
+import os
+import re
+import math
+import typing as t
+import logging
+import functools
+from abc import ABCMeta
+from abc import abstractmethod
+
+import psutil
+
+from ..exceptions import BentoMLConfigException
+
+logger = logging.getLogger(__name__)
+
+_RESOURCE_REGISTRY: dict[str, t.Type[Resource[t.Any]]] = {}
+
+T = t.TypeVar("T")
+
+
+def get_resource(resources: dict[str, t.Any], resource_kind: str) -> t.Any:
+    if resource_kind not in _RESOURCE_REGISTRY:
+        raise BentoMLConfigException(f"Unknown resource kind '{resource_kind}'.")
+
+    resource: t.Type[Resource[t.Any]] = _RESOURCE_REGISTRY[resource_kind]
+
+    if resource_kind in resources:
+        if resources[resource_kind] == "system":
+            return resource.from_system()
+        else:
+            return resource.from_spec(resources[resource_kind])
+    else:
+        return None
+
+
+def system_resources() -> dict[str, t.Any]:
+    res: dict[str, t.Any] = {}
+    for resource_kind, resource in _RESOURCE_REGISTRY.items():
+        res[resource_kind] = resource.from_system()
+    return res
+
+
+class ResourceMeta(ABCMeta):
+    def __new__(  # pylint: disable=arguments-differ # we don't want to be generic
+        cls,
+        name: str,
+        bases: tuple[type],
+        attr_dict: dict[t.Any, t.Any],
+        *,
+        resource_id: str,
+    ):
+        res = super().__new__(cls, name, bases, attr_dict)
+        if resource_id != "":
+            # register the class with RESOURCE_REGISTRY
+            assert issubclass(res, Resource)
+            _RESOURCE_REGISTRY[resource_id] = res
+
+        return res  # pylint: disable=useless-super-delegation (???)
+
+
+class Resource(t.Generic[T], metaclass=ResourceMeta, resource_id=""):
+    @classmethod
+    @abstractmethod
+    def from_spec(cls, spec: t.Any) -> T:
+        """
+        Get an instance of this resource from user input. For example, a CPU resource
+        might parse "10m" and return a CPU resource with 0.01 CPUs.
+        """
+
+    @classmethod
+    @abstractmethod
+    def from_system(cls) -> T:
```

Comment:
```suggestion
    def from_system(cls) -> "t.TypeGuard[T]":
```

Reply:
Hm, is this a type guard? Aren't type guards supposed to return booleans?

Reply:
type guards ensure the return type is the correct type, not a boolean. Essentially it does type narrowing for us without having to implement `if isinstance`

Reply:
https://peps.python.org/pep-0647/

Reply:
"...Return statements within a type guard function should return bool values, and type checkers should verify that all return paths return a bool."

I'm pretty sure that `TypeGuard` is mostly for implementing custom `isinstance`.

Reply:
noted.